File: app/src/main/java/net/bloople/stories/IndexViewModel.java
package net.bloople.stories;

import android.app.Application;
import android.database.Cursor;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class IndexViewModel extends AndroidViewModel {
    private Application application;
    private MutableLiveData<Cursor> searchResults;
    private BooksSearcher searcher = new BooksSearcher();
    private MutableLiveData<String> sorterDescription;

    public IndexViewModel(@NonNull Application application) {
        super(application);
        this.application = application;
    }

    public LiveData<Cursor> getSearchResults() {
        if(searchResults == null) {
            searchResults = new MutableLiveData<>();
            resolve();
        }
        return searchResults;
    }

    public LiveData<String> getSorterDescription() {
        if(sorterDescription == null) {
            sorterDescription = new MutableLiveData<>(searcher.description());
        }
        return sorterDescription;
    }

    public int getSortMethod() {
        return searcher.getSortMethod();
    }

    public boolean getSortDirectionAsc() {
        return searcher.getSortDirectionAsc();
    }

    public void setSearchText(String searchText) {
        searcher.setSearchText(searchText);
        resolve();
    }

    public void setSort(int sortMethod, boolean sortDirectionAsc) {
        searcher.setSortMethod(sortMethod);
        searcher.setSortDirectionAsc(sortDirectionAsc);
        sorterDescription.setValue(searcher.description());
        resolve();
    }

    private void resolve() {
        ExecutorService service =  Executors.newSingleThreadExecutor();
        service.submit(() -> {
            searchResults.postValue(searcher.search(application));
        });
    }
}

File: app/src/main/java/net/bloople/stories/IndexViewModel.kt
package net.bloople.stories

import android.app.Application
import android.database.Cursor
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.MutableLiveData
import java.util.concurrent.Executors

class IndexViewModel(application: Application) : AndroidViewModel(application) {
    val searchResults: MutableLiveData<Cursor> by lazy {
        MutableLiveData<Cursor>().also { resolve() }
    }

    val sorterDescription: MutableLiveData<String> by lazy {
        MutableLiveData<String>(searcher.description())
    }

    private val searcher = BooksSearcher()

    val sortMethod: Int
        get() = searcher.sortMethod

    val sortDirectionAsc: Boolean
        get() = searcher.sortDirectionAsc

    fun setSearchText(searchText: String?) {
        searcher.setSearchText(searchText)
        resolve()
    }

    fun setSort(sortMethod: Int, sortDirectionAsc: Boolean) {
        searcher.sortMethod = sortMethod
        searcher.sortDirectionAsc = sortDirectionAsc
        sorterDescription.value = searcher.description()
        resolve()
    }

    fun refresh() {
        resolve()
    }

    private fun resolve() {
        val service = Executors.newSingleThreadExecutor()
        service.submit { searchResults.postValue(searcher.search(getApplication())) }
    }
}